pkg/dnsclient/dns.go

```diff
@@ -0,0 +1,347 @@
+		if resp != nil {
+			result = dns.RcodeToString[resp.Rcode]
+		}
+
+		metrics.MeasureSince(
+			metricsKeyForTotalLookup,
+			started,
+			metrics.Tag{Name: "qtype", Value: qtypeStr},
+			metrics.Tag{Name: "result", Value: result},
+			metrics.Tag{Name: "retries", Value: strconv.Itoa(tries)},
+			metrics.Tag{Name: "resolver", Value: chosenServer},
+		)
+	}()
+
+	type dnsResp struct {
+		m   *dns.Msg
+		err error
+	}
+
+	for {
+		ch := make(chan dnsResp, 1)
+
+		go func() {
+			logger.Tracef("hostname=%q, type=%v, server=%s", hostname, qtypeStr, chosenServer)
+
+			started := time.Now().UTC()
+			rsp, _, err := c.dnsClient.Exchange(m, chosenServer)
+			result := "failed"
+			if rsp != nil {
+				result = dns.RcodeToString[rsp.Rcode]
+			}
+			metrics.MeasureSince(
+				metricsKeyForQuery,
+				started,
+				metrics.Tag{Name: "qtype", Value: qtypeStr},
+				metrics.Tag{Name: "result", Value: result},
+				metrics.Tag{Name: "resolver", Value: chosenServer},
+			)
+
+			ch <- dnsResp{m: rsp, err: err}
+		}()
+		select {
+		case <-ctx.Done():
+			reason := "unknown"
+			if ctx.Err() == context.DeadlineExceeded {
+				reason = "deadline"
+			} else if ctx.Err() == context.Canceled {
+				reason = "canceled"
+			}
+			metrics.IncrCounter(
+				metricsKeyForTimeoutCounter,
+				1,
+				metrics.Tag{Name: "qtype", Value: qtypeStr},
+				metrics.Tag{Name: "reason", Value: reason},
+				metrics.Tag{Name: "resolver", Value: chosenServer},
+			)
+
+			err = ctx.Err()
+			logger.Errorf("hostname=%q, type=%v, err=[%s]", hostname, qtypeStr, err.Error())
+			return
+		case r := <-ch:
+			if r.err != nil {
+				operr, ok := r.err.(*net.OpError)
+				isRetryable := ok && operr.Temporary()
+				hasRetriesLeft := tries < c.maxTries
+				if isRetryable && hasRetriesLeft {
+					tries++
+					// Chose a new server to retry the query with by incrementing the
+					// chosen server index modulo the number of servers. This ensures that
+					// if one dns server isn't available we retry with the next in the
+					// list.
+					chosenServerIndex = (chosenServerIndex + 1) % len(c.servers)
+					chosenServer = c.servers[chosenServerIndex]
+					continue
+				} else if isRetryable && !hasRetriesLeft {
+					logger.Errorf("reason=out_of_retries, hostname=%q, type=%v", hostname, qtypeStr)
+				}
+			}
+			resp, err = r.m, r.err
+			return
+		}
+	}
+}
+
+// LookupTXT sends a DNS query to find all TXT records associated with
+// the provided hostname which it returns along with the returned
+// DNS authority section.
+func (c *Client) LookupTXT(ctx context.Context, hostname string) ([]string, []string, error) {
+	var txt []string
+	dnsType := dns.TypeTXT
+	r, err := c.exchangeOne(ctx, hostname, dnsType)
+	if err != nil {
+		logger.Tracef("reason=exchangeOne, host=%s, err=[%s]", hostname, err.Error())
+		return nil, nil, &Error{dnsType, hostname, err, -1}
+	}
+	if r.Rcode != dns.RcodeSuccess {
+		logger.Tracef("reason=exchangeOne, host=%s, rc=%d", hostname, r.Rcode)
```

Comment:
Tracef or Errorf?

Reply:
Trace, because the caller should log error if needed